src/libraries/System.Net.Quic/src/System/Net/Quic/QuicStream.cs

```diff
@@ -3,140 +3,601 @@
         {
-            ValidateBufferArguments(buffer, offset, count);
-            return Read(buffer.AsSpan(offset, count));
+            throw new InvalidOperationException(SR.net_quic_reading_notallowed);
         }
 
-        public override int ReadByte()
+        if (NetEventSource.Log.IsEnabled())
         {
-            byte b = 0;
-            return Read(new Span<byte>(ref b)) != 0 ? b : -1;
+            NetEventSource.Info(this, $"{this} Stream reading into memory of '{buffer.Length}' bytes.");
         }
 
-        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        if (_receiveTcs.IsCompleted)
         {
-            ValidateBufferArguments(buffer, offset, count);
-            return ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();
+            // Special case exception type for pre-canceled token while we've already transitioned to a final state and don't need to abort read.
+            // It must happen before we try to get the value task, since the task source is versioned and each instance must be awaited.
+            cancellationToken.ThrowIfCancellationRequested();
         }
 
-        public override void Write(byte[] buffer, int offset, int count)
+        // The following loop will repeat at most twice depending whether some data are readily available in the buffer (one iteration) or not.
+        // In which case, it'll wait on RECEIVE or any of PEER_SEND_(SHUTDOWN|ABORTED) event and attempt to copy data in the second iteration.
+        int totalCopied = 0;
+        do
         {
-            ValidateBufferArguments(buffer, offset, count);
-            Write(buffer.AsSpan(offset, count));
+            // Concurrent call, this one lost the race.
+            if (!_receiveTcs.TryGetValueTask(out ValueTask valueTask, this, cancellationToken))
+            {
+                throw new InvalidOperationException(SR.Format(SR.net_io_invalidnestedcall, "read"));
+            }
+
+            // Copy data from the buffer, reduce target and increment total.
+            int copied = _receiveBuffers.CopyTo(buffer, out bool complete, out bool empty);
+            buffer = buffer.Slice(copied);
+            totalCopied += copied;
+
+            // Make sure the task transitions into final state before the method finishes.
+            if (complete)
+            {
+                _receiveTcs.TrySetResult(final: true);
+            }
+
+            // Unblock the next await to end immediately, i.e. there were/are any data in the buffer.
+            if (totalCopied > 0 || !empty)
+            {
+                _receiveTcs.TrySetResult();
+            }
+
+            // This will either wait for RECEIVE event (no data in buffer) or complete immediately and reset the task.
+            await valueTask.ConfigureAwait(false);
+
+            // This is the last read, finish even despite not copying anything.
+            if (complete)
```

Comment:
Do I understand correctly, that `complete` can *only* become true in "synchronous" case? Meaning, we didn't actually wait for anything in valueTask above